src/lib/repositories/impl/inventory_ingredient_repository_impl.py

```diff
@@ -69,3 +74,25 @@ def validate_ingredient_availability(
         if ingredient_to_validate[0].ingredient_quantity > quantity_to_use:
             return True
         return False
+
+    def get_final_product_qty_by_product_ids(self, product_ids):
+
+        final_product_possible_qty_map = reduce(
```

Comment:
Try to do something like:
```
reducer1 = reduce(...)
reducer2 = reduce(..., reducer1)
```

This is just an idea. Just avoid nested code.

Reply:
done